src/main/scala/firrtl/transforms/InlineCasts.scala

```diff
@@ -28,23 +28,30 @@ object InlineCastsTransform {
     * @param expr the Expression being transformed
     * @return Returns expr with [[WRef]]s replaced by values found in replace
     */
-  def onExpr(replace: NodeMap)(expr: Expression): Expression = expr match {
-    // Anything that may generate a part-select should not be inlined!
-    case DoPrim(op, _, _, _) if (isBitExtract(op) || op == Pad) => expr
-    case e =>
-      e.map(onExpr(replace)) match {
-        case e @ WRef(name, _, _, _) =>
-          replace
-            .get(name)
-            .filter(isSimpleCast(castSeen = false))
-            .getOrElse(e)
-        case e @ DoPrim(op, Seq(WRef(name, _, _, _)), _, _) if isCast(op) =>
-          replace
-            .get(name)
-            .map(value => e.copy(args = Seq(value)))
-            .getOrElse(e)
-        case other => other // Not a candidate
-      }
+  def onExpr(replace: NodeMap)(expr: Expression): Expression = {
+    // Keep track if we've seen any non-cast expressions while recursing
+    def rec(seenComplexExpr: Boolean)(expr: Expression): Expression = expr match {
+      // Anything that may generate a part-select should not be inlined!
+      case DoPrim(op, _, _, _) if (isBitExtract(op) || op == Pad) => expr
```

Comment:
Does this restriction actually serve any purpose? I imagine it might have just prevented a particularly common degenerate case of the bug this PR fixes.

Reply:
Originally I thought this was a degenerate case but then reasoned myself into thinking it wasn't. Now I agree with you again so I'm going to try deleting this.

Reply:
Alright so this is still required because basically we cannot emit a pad of a literal. This is why we have `legalizePad` which doesn't support cast literals at the moment: https://github.com/chipsalliance/firrtl/blob/94d1bee4c23bd3d8f99dae3ca431ffaa5dc1410d/src/main/scala/firrtl/passes/Legalize.scala#L52

I think the solution is to remove this special case but add `legalizeCast` to ensure that cast literals are always converted.

Reply:
Maybe update the comment to say that you skip pads to avoid inlining literals into pads.

Reply:
Leaving the suggestion above to future work.